File: backend/services/report_service.py
"""User-generated content reports service."""
from __future__ import annotations
from datetime import datetime, timezone
from bson import ObjectId
from fastapi import HTTPException

from database import get_db

VALID_TARGETS = {"listing", "user", "review", "message"}
VALID_REASONS = {"spam", "offensive", "scam", "wrong_category", "other"}
VALID_ACTIONS = {"takedown", "warn", "ban", "none"}


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _s(d: dict) -> dict:
    out = dict(d)
    _id = out.pop("_id", None)
    if _id is not None:
        out["id"] = str(_id)
    return out


async def create_report(reporter_id: str, body: dict) -> dict:
    if body.get("target_type") not in VALID_TARGETS:
        raise HTTPException(400, f"target_type must be in {sorted(VALID_TARGETS)}")
    if body.get("reason") not in VALID_REASONS:
        raise HTTPException(400, f"reason must be in {sorted(VALID_REASONS)}")
    target_id = body.get("target_id", "")
    # Loose validation — for listing/user/review target_id must be an ObjectId
    if body["target_type"] in ("listing", "user", "review", "message"):
        if not ObjectId.is_valid(target_id):
            raise HTTPException(400, "target_id must be a valid id")
    db = get_db()
    doc = {
        "reporter_id": reporter_id,
        "target_type": body["target_type"],
        "target_id": target_id,
        "reason": body["reason"],
        "description": (body.get("description") or "")[:1000],
        "status": "open",
        "resolved_by": None, "resolved_at": None,
        "resolution_action": None, "resolution_note": None,
        "created_at": _now(), "updated_at": _now(),
    }
    res = await db.reports.insert_one(doc)
    doc["_id"] = res.inserted_id
    return _s(doc)


async def list_reports(status: str | None, target_type: str | None, cursor: str | None, limit: int) -> dict:
    db = get_db()
    q: dict = {}
    if status:
        q["status"] = status
    if target_type:
        q["target_type"] = target_type
    if cursor and ObjectId.is_valid(cursor):
        q["_id"] = {"$lt": ObjectId(cursor)}
    docs = await db.reports.find(q).sort([("_id", -1)]).limit(limit + 1).to_list(limit + 1)
    has_more = len(docs) > limit
    docs = docs[:limit]
    return {
        "items": [_s(d) for d in docs],
        "next_cursor": str(docs[-1]["_id"]) if has_more and docs else None,
        "has_more": has_more,
    }


async def resolve_report(rid: str, admin_id: str, action: str, note: str | None) -> dict:
    if action not in VALID_ACTIONS:
        raise HTTPException(400, f"action must be in {sorted(VALID_ACTIONS)}")
    db = get_db()
    if not ObjectId.is_valid(rid):
        raise HTTPException(400, "Invalid report id")
    rep = await db.reports.find_one({"_id": ObjectId(rid)})
    if not rep:
        raise HTTPException(404, "Report not found")

    # Apply action side-effects
    if action == "takedown" and rep["target_type"] == "listing" and ObjectId.is_valid(rep["target_id"]):
        await db.listings.update_one(
            {"_id": ObjectId(rep["target_id"])},
            {"$set": {"is_takendown": True, "status": "paused", "updated_at": _now()}},
        )
    elif action == "ban" and rep["target_type"] == "user" and ObjectId.is_valid(rep["target_id"]):
        await db.users.update_one(
            {"_id": ObjectId(rep["target_id"])},
            {"$set": {"is_banned": True, "is_active": False, "updated_at": _now()}},
        )

    await db.reports.update_one(
        {"_id": ObjectId(rid)},
        {"$set": {
            "status": "resolved", "resolved_by": admin_id, "resolved_at": _now(),
            "resolution_action": action, "resolution_note": note, "updated_at": _now(),
        }},
    )
    return _s(await db.reports.find_one({"_id": ObjectId(rid)}))


async def dismiss_report(rid: str, admin_id: str, reason: str | None) -> dict:
    db = get_db()
    if not ObjectId.is_valid(rid):
        raise HTTPException(400, "Invalid report id")
    rep = await db.reports.find_one({"_id": ObjectId(rid)})
    if not rep:
        raise HTTPException(404, "Report not found")
    await db.reports.update_one(
        {"_id": ObjectId(rid)},
        {"$set": {
            "status": "dismissed", "resolved_by": admin_id, "resolved_at": _now(),
            "resolution_action": "none", "resolution_note": reason, "updated_at": _now(),
        }},
    )
    return _s(await db.reports.find_one({"_id": ObjectId(rid)}))


async def open_count() -> int:
    db = get_db()
    return await db.reports.count_documents({"status": "open"})

File: backend/src/services/report.service.js
const { Report } = require('../models/Misc');
const Listing = require('../models/Listing');
const User = require('../models/User');
const { Review } = require('../models/Phase4');
const ApiError = require('../utils/ApiError');

const VALID_TARGETS = new Set(['listing', 'user', 'review', 'message']);
const VALID_REASONS = new Set(['spam', 'offensive', 'scam', 'wrong_category', 'other']);
const VALID_ACTIONS = new Set(['takedown', 'warn', 'ban', 'none']);

const serializeReport = (d) => {
  if (!d) return null;
  const out = d.toObject ? d.toObject() : { ...d };
  if (out._id) {
    out.id = out._id.toString();
    delete out._id;
  }
  return out;
};

const createReport = async (reporterId, body) => {
  if (!VALID_TARGETS.has(body.target_type)) {
    throw ApiError.badRequest(`target_type must be in ${Array.from(VALID_TARGETS).sort().join(', ')}`);
  }
  if (!VALID_REASONS.has(body.reason)) {
    throw ApiError.badRequest(`reason must be in ${Array.from(VALID_REASONS).sort().join(', ')}`);
  }
  const targetId = body.target_id || '';
  if (['listing', 'user', 'review', 'message'].includes(body.target_type)) {
    if (!targetId) {
      throw ApiError.badRequest('target_id must be a valid id');
    }
  }

  const doc = await Report.create({
    reporter_id: reporterId,
    target_type: body.target_type,
    target_id: targetId,
    reason: body.reason,
    description: String(body.description || '').slice(0, 1000),
    status: 'open',
    resolved_by: null,
    resolved_at: null,
    resolution_action: null,
    resolution_note: null,
  });

  return serializeReport(doc);
};

const hydrateTargets = async (items) => {
  const listingIds = [];
  const userIds = [];
  const reviewIds = [];

  for (const r of items) {
    const tid = r.target_id;
    if (!tid) continue;
    if (r.target_type === 'listing') {
      listingIds.push(tid);
    } else if (r.target_type === 'user') {
      userIds.push(tid);
    } else if (r.target_type === 'review') {
      reviewIds.push(tid);
    }
  }

  const listingsMap = {};
  if (listingIds.length > 0) {
    const listings = await Listing.find({ _id: { $in: listingIds } }, { title: 1, slug: 1, is_takendown: 1 });
    for (const li of listings) {
      listingsMap[li._id.toString()] = li;
    }
  }

  const usersMap = {};
  if (userIds.length > 0) {
    const users = await User.find({ _id: { $in: userIds } }, { name: 1, phone: 1, is_banned: 1 });
    for (const u of users) {
      usersMap[u._id.toString()] = u;
    }
  }

  const reviewsMap = {};
  if (reviewIds.length > 0) {
    const reviews = await Review.find({ _id: { $in: reviewIds } }, { rating: 1, comment: 1 });
    for (const rv of reviews) {
      reviewsMap[rv._id.toString()] = rv;
    }
  }

  for (const r of items) {
    const tid = r.target_id;
    r.target_label = null;
    r.target_link = null;

    if (r.target_type === 'listing' && listingsMap[tid]) {
      const li = listingsMap[tid];
      r.target_label = li.title || '(untitled listing)';
      r.target_link = `/listing/${li.slug || tid}`;
      r.target_status = li.is_takendown ? 'taken_down' : 'active';
    } else if (r.target_type === 'user' && usersMap[tid]) {
      const u = usersMap[tid];
      r.target_label = u.name || u.phone || '(unnamed user)';
      r.target_link = `/vendor/${tid}`;
      r.target_status = u.is_banned ? 'banned' : 'active';
    } else if (r.target_type === 'review' && reviewsMap[tid]) {
      const rv = reviewsMap[tid];
      r.target_label = `${rv.rating || '?'}★ · ${(rv.comment || '').slice(0, 60)}`;
    }
  }

  return items;
};

const listReports = async (status = null, targetType = null, cursor = null, limit = 20) => {
  const q = {};
  if (status) {
    q.status = status;
  }
  if (targetType) {
    q.target_type = targetType;
  }
  if (cursor) {
    q._id = { $lt: cursor };
  }

  const docs = await Report.find(q).sort({ _id: -1 }).limit(limit + 1);
  const hasMore = docs.length > limit;
  const sliced = docs.slice(0, limit);
  const items = sliced.map(serializeReport);

  const hydrated = await hydrateTargets(items);

  return {
    items: hydrated,
    next_cursor: hasMore && hydrated.length > 0 ? hydrated[hydrated.length - 1].id : null,
    has_more: hasMore,
  };
};

const resolveReport = async (rid, adminId, action, note) => {
  if (!VALID_ACTIONS.has(action)) {
    throw ApiError.badRequest(`action must be in ${Array.from(VALID_ACTIONS).sort().join(', ')}`);
  }

  const rep = await Report.findById(rid);
  if (!rep) {
    throw ApiError.notFound('Report not found');
  }

  const now = new Date().toISOString();

  // Apply action side-effects
  if (action === 'takedown' && rep.target_type === 'listing') {
    await Listing.updateOne(
      { _id: rep.target_id },
      { $set: { is_takendown: true, status: 'paused', updated_at: now } }
    );
  } else if (action === 'ban' && rep.target_type === 'user') {
    await User.updateOne(
      { _id: rep.target_id },
      { $set: { is_banned: true, is_active: false, updated_at: now } }
    );
  }

  const updated = await Report.findOneAndUpdate(
    { _id: rid },
    {
      $set: {
        status: 'resolved',
        resolved_by: adminId,
        resolved_at: now,
        resolution_action: action,
        resolution_note: note,
        updated_at: now,
      },
    },
    { returnDocument: 'after' }
  );

  return serializeReport(updated);
};

const dismissReport = async (rid, adminId, reason = null) => {
  const rep = await Report.findById(rid);
  if (!rep) {
    throw ApiError.notFound('Report not found');
  }

  const now = new Date().toISOString();
  const updated = await Report.findOneAndUpdate(
    { _id: rid },
    {
      $set: {
        status: 'dismissed',
        resolved_by: adminId,
        resolved_at: now,
        resolution_action: 'none',
        resolution_note: reason,
        updated_at: now,
      },
    },
    { returnDocument: 'after' }
  );

  return serializeReport(updated);
};

const openCount = async () => {
  return await Report.countDocuments({ status: 'open' });
};

module.exports = {
  createReport,
  listReports,
  resolveReport,
  dismissReport,
  openCount,
};
